# Backport of upstream PR #777 (fixes issue #767 for v2.10.3)
# https://github.com/sahib/rmlint/pull/777
# Port command line parsing to Python standard library argparse as PyGObject >= 3.56 removed GLib.option.

diff --git a/gui/shredder/cmdline.py b/gui/shredder/cmdline.py
index 54ec6d353..a8134f67b 100644
--- a/gui/shredder/cmdline.py
+++ b/gui/shredder/cmdline.py
@@ -8,13 +8,13 @@
 """
 
 # Stdlib:
+import argparse
 import os
 import sys
 import logging
 
 # External:
 from gi.repository import Gio
-from gi.repository import GLib
 
 
 def show_version():
@@ -45,84 +45,86 @@ def adjust_loglevel(root_logger, count):
 
 
 def parse_arguments(root_logger):
-    """Parse the cmdline options using GOption."""
+    """Parse the cmdline options using argparse."""
     sys.argv[0] = 'shredder'
-    parser = GLib.option.OptionParser(
-        "PATHS ...",
+    parser = argparse.ArgumentParser(
+        prog='shredder',
         description="A gui frontend to rmlint.",
-        option_list=[
-            GLib.option.make_option(
-                "--add-location", "-a",
-                type="filename",
-                action="append",
-                dest="locations",
-                help="Add locations to locations view."
-            ),
-            GLib.option.make_option(
-                "--scan", "-s",
-                type="filename",
-                action="append",
-                dest="untagged",
-                help="Add location to scan (as untagged path)."
-            ),
-            GLib.option.make_option(
-                "--scan-tagged", "-S",
-                type="filename",
-                action="append",
-                dest="tagged",
-                help="Add location to scan (as tagged path)."
-            ),
-            GLib.option.make_option(
-                "--load-script", "-l",
-                type="filename",
-                action="store",
-                dest="script",
-                help="Show `script` in editor view."
-            ),
-            GLib.option.make_option(
-                "--verbose", "-v",
-                action="count",
-                dest='more_verbosity',
-                help="Be more verbose."
-            ),
-            GLib.option.make_option(
-                "--less-verbose", "-V",
-                action="count",
-                dest='less_verbosity',
-                help="Be less verbose."
-            ),
-            GLib.option.make_option(
-                "--show-settings", "-c",
-                action="store_true",
-                dest='show_settings',
-                help="Show the settings view."
-            ),
-            GLib.option.make_option(
-                "--version", "",
-                action="store_true",
-                dest="show_version",
-                help="Show the version of Shredder."
-            ),
-        ]
+    )
+    parser.add_argument(
+        "-a", "--add-location",
+        action="append",
+        dest="locations",
+        metavar="PATH",
+        help="Add locations to locations view."
+    )
+    parser.add_argument(
+        "-s", "--scan",
+        action="append",
+        dest="untagged",
+        metavar="PATH",
+        help="Add location to scan (as untagged path)."
+    )
+    parser.add_argument(
+        "-S", "--scan-tagged",
+        action="append",
+        dest="tagged",
+        metavar="PATH",
+        help="Add location to scan (as tagged path)."
+    )
+    parser.add_argument(
+        "-l", "--load-script",
+        dest="script",
+        help="Show `script` in editor view."
+    )
+    parser.add_argument(
+        "-v", "--verbose",
+        action="count",
+        dest='more_verbosity',
+        help="Be more verbose."
+    )
+    parser.add_argument(
+        "-V", "--less-verbose",
+        action="count",
+        dest='less_verbosity',
+        help="Be less verbose."
+    )
+    parser.add_argument(
+        "-c", "--show-settings",
+        action="store_true",
+        dest='show_settings',
+        help="Show the settings view."
+    )
+    parser.add_argument(
+        "--version",
+        action="store_true",
+        dest="show_version",
+        help="Show the version of Shredder."
+    )
+    parser.add_argument(
+        "paths",
+        nargs="*",
+        metavar="PATH",
+        help=argparse.SUPPRESS
     )
 
-    try:
-        parser.parse_args()
-    except GLib.Error as err:
-        root_logger.error(str(err))
-        return None
-
-    vals = parser.values
-    if parser.values.show_version:
+    vals = parser.parse_args()
+    if vals.show_version:
         show_version()
 
     adjust_loglevel(
         root_logger,
-        (vals.more_verbosity or 0) +
+        (vals.more_verbosity or 0) -
         (vals.less_verbosity or 0) +
         4  # Default loglevel: info.
     )
 
+    for path in vals.paths:
+        root_logger.warning(
+            'argument %s ignored. Will be rejected in a future version.',
+            path
+        )
+
     # Check paths to be valid:
     paths = (vals.tagged or []) + (vals.untagged or []) + [vals.script]
     for path in (path for path in paths if path):
